src/libs/type-utils/lodash-fp-helpers.ts

```diff
@@ -0,0 +1,118 @@
+ * handling logic which still honors the same function signature of the fn
+ * it is given as it's single argument
+ * @param handlers array of handler (wrapper) functions to apply to mainFn
+ * @param mainFn the main function being wrapped with handlers
+ */
+export const withHandlers : WithHandlersFn = <F extends AnyFn>(handlers: WrapFn<F>[], mainFn: F) => {
+  const resultFn = _.flow(...handlers)(mainFn)
+  return resultFn
+}
+
+export interface CurryLastArgFn {
+  <A, LAST, R>(
+      fn: (a: A, last: LAST) => R
+  ) : (a: A) => (last: LAST) => R;
+  <A, B, LAST, R>(
+      fn: (a: A, b: B, last: LAST) => R
+  ) : (a: A, b: B) => (last: LAST) => R;
+  <A, B, C, LAST, R>(
+      fn: (a: A, b: B, c: C, last: LAST) => R
+  ) : (a: A, b: B, c: C) => (last: LAST) => R;
+  <A, B, C, D, LAST, R>(
+      fn: (a: A, b: B, c: C, d: D, last: LAST) => R
+  ) : (a: A, b: B, c: C, d: D) => (last: LAST) => R;
+}
+
+/**
+ * produces a curried function that expect the last argument to be given
+ * before executing the original function.
+ * Provides better type flow then the more general _.curry(fn)
+ * @param fn function with up to 5 arguments
+ */
+export const curryLastArg: CurryLastArgFn = (fn: (
+    ...args: unknown[]) => unknown) => {
+  return (...args2: unknown[]) => {
+    return (last: unknown) => {
+      return fn(...args2, last)
+    }
+  }
+}
+
+/**
+ * (non-async) a convenience helper for creating handler-wrapper functions
+ * that add logic and/or side effects but produce a wrapped function that
+ * preserves the function signature of the original function.  This helper
+ * removes the concern of original function arguments by providing a
+ * zero-argument executor argument that the given handler function can call.
+ * When used with recommended patterns (see unit tests) the need for
+ * currying is removed.
+ * @param handler
+ */
+export const createHandler = <F extends AnyFn>(
+  handler: (executor: () => ReturnType<F>) => ReturnType<F>
+) => {
+  const wrappedFn = (fn: (...args: Parameters<F>) => ReturnType<F>) => {
+    const innerFn = (...fnArgs: Parameters<F>): ReturnType<F> => {
+      const executor = (): ReturnType<F> => {
+        const result: ReturnType<F> = fn(...fnArgs)
+        return result
+      }
+      const handlerResult = handler(executor)
+      return handlerResult
+    }
+    return innerFn
+  }
+  return wrappedFn
+}
+
+/**
+ * (async) a convenience helper for creating handler-wrapper functions
+ * that add logic and/or side effects but produce a wrapped function that
+ * preserves the function signature of the original function.  This helper
+ * removes the concern of original function arguments by providing a
+ * zero-argument executor argument that the given handler function can call.
+ * When used with recommended patterns (see unit tests) the need for
+ * currying is removed.
+ * @param handler
+ */
+export const createHandlerAsync = <P, F extends AnyPromiseFn<P>>(
```

Comment:
I toyed around with the overload route, but as I suspected, async comes with distinct expectations that are different from non-asnyc as far as execution order.

createHandlerAsyc expects everything to be async, including the handler fn being passed in.  So everything is properly awaited to match execution order expectations of async.

Reply:
I still don't see how `createAsyncHandler<P, () => Promise<P>>` differs from `createHandler<Promise<P>>` in anything but the type signature.

Reply:
execution order.  Calling a method that returns a promise results in a fire-and-forget flow where the caller immediately resumes execution as though the call was instant, but it is off in the background eventually resolving a dangling promise.

While this article is from a related topic, it demonstrates the same root concern:

https://stackoverflow.com/questions/55262996/does-awaiting-a-non-promise-have-any-detectable-effect

Reply:
I'm still skeptical. I don't see a dangling promise here because `result` and `handlerResult` are returned. If `handler` doesn't await `executor` or whatever uses the created the handler doesn't await it, that's on them. It still seems like the difference is the type signatures. I could be wrong though. And in any case, I'm not opposed to a separate function for async handlers, so I won't argue further.

Reply:
I think your right, but I had some difficulties wrangling the types until I separated them out... I would like to give it another go, but as a follow-up PR to give space for surprises.  I think some of the other type flow improvements may make this easier to combine into an overload-typed single function.